### PR TITLE
rootless, overlay: use fuse-overlayfs

### PR DESCRIPTION
**Pull request by giuseppe**
when running in rootless mode, use fuse-overlayfs for mounting the
overlay file system on the host.  Then create a bind mount inside the
container.

Closes: https://github.com/containers/buildah/issues/1741

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

**Comment by giuseppe** [comment]
marked as WIP, as it should honor the `mount_program` option in the storage configuration file

**Comment by giuseppe** [comment]
> marked as WIP, as it should honor the `mount_program` option in the storage configuration file

fixed now

File: pkg/overlay/overlay.go [hunk 1]
```diff
@@ -46,24 +49,82 @@ func MountTemp(store storage.Store, containerId, source, dest string, rootUID, r
...
 // CleanupContent removes all temporary mountpoint and all content from
 // directory
 func CleanupContent(containerDir string) (Err error) {
 	contentDir := filepath.Join(containerDir, "overlay")
+
+	if unshare.IsRootless() {
+		mergeDir := filepath.Join(contentDir, "merge")
+		if err := unix.Unmount(mergeDir, 0); err != nil {
+			if !os.IsNotExist(err) {
+				return errors.Wrapf(err, "unmount overlay")
```
Comment (rhatdan) [inline pkg/overlay/overlay.go]: Add Path to error.

File: pkg/overlay/overlay.go [hunk 2]
```diff
@@ -46,24 +49,82 @@ func MountTemp(store storage.Store, containerId, source, dest string, rootUID, r
...
+			s := strings.SplitN(i, "=", 2)
+			if len(s) != 2 {
+				continue
+			}
+			k := s[0]
+			v := s[1]
+			switch k {
+			case ".mount_program", "overlay.mount_program", "overlay2.mount_program":
+				mountProgram = v
+				break loop
```
Comment (rhatdan) [inline pkg/overlay/overlay.go]: Why do you need loop: and break loop?  Wouldn't break just work?

File: pkg/overlay/overlay.go [hunk 3]
```diff
@@ -46,24 +49,82 @@ func MountTemp(store storage.Store, containerId, source, dest string, rootUID, r
...
+			mergeDir := filepath.Join(contentDir, "merge")
+
+			if err := idtools.MkdirAllAs(mergeDir, 0700, rootUID, rootGID); err != nil {
+				return mount, "", errors.Wrapf(err, "failed to create the overlay %s directory", mergeDir)
+			}
+
+			cmd := exec.Command(mountProgram, "-o", overlayOptions, mergeDir)
+
+			if err := cmd.Run(); err != nil {
+				return mount, "", errors.Wrapf(err, "exec fuse-overlayfs")
```
Comment (rhatdan) [inline pkg/overlay/overlay.go]: Should be mountProgram not fuse-overlayfs.

**Comment by rhatdan** [comment]
Do we have tests for this that will run in rootless mode?

File: pkg/overlay/overlay.go (hunk 2 above)
Comment (giuseppe) [reply pkg/overlay/overlay.go]: writing too much C recently.  I assumed it would be part of the switch.

File: pkg/overlay/overlay.go [hunk 4]
```diff
@@ -46,24 +49,81 @@ func MountTemp(store storage.Store, containerId, source, dest string, rootUID, r
...
+		mountProgram := ""
+
+		for _, i := range store.GraphOptions() {
+			s := strings.SplitN(i, "=", 2)
+			if len(s) != 2 {
+				continue
+			}
+			k := s[0]
+			v := s[1]
+			switch k {
```
Comment (rhatdan) [inline pkg/overlay/overlay.go]: How about you setup a map
mountMap := map[string]bool {
               ".mount_program": true,
               "overlay.mount_program": true,
               "overlay2.mount_program":true, 
}
...
if mountMap[k] {
           mountProgram=v
           break
}

Comment (giuseppe) [reply pkg/overlay/overlay.go]: amended in the last version

**Comment by rhatdan** [comment]
LGTM
@vrothberg @baude @mheon @TomSweeneyRedHat @QiWang19 PTAL

**Comment by TomSweeneyRedHat** [comment]
LGTM

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**